compiler/rustc_session/src/config.rs

```diff
@@ -232,75 +233,50 @@ pub struct LinkSelfContained {
     /// Used for compatibility with the existing opt-in and target inference.
     pub explicitly_set: Option<bool>,
 
-    /// The components that are enabled.
-    components: LinkSelfContainedComponents,
-}
+    /// The components that are enabled on the CLI, using the `+component` syntax or one of the
+    /// `true` shorcuts.
+    enabled_components: LinkSelfContainedComponents,
 
-bitflags::bitflags! {
-    #[derive(Default)]
-    /// The `-C link-self-contained` components that can individually be enabled or disabled.
-    pub struct LinkSelfContainedComponents: u8 {
-        /// CRT objects (e.g. on `windows-gnu`, `musl`, `wasi` targets)
-        const CRT_OBJECTS = 1 << 0;
-        /// libc static library (e.g. on `musl`, `wasi` targets)
-        const LIBC        = 1 << 1;
-        /// libgcc/libunwind (e.g. on `windows-gnu`, `fuchsia`, `fortanix`, `gnullvm` targets)
-        const UNWIND      = 1 << 2;
-        /// Linker, dlltool, and their necessary libraries (e.g. on `windows-gnu` and for `rust-lld`)
-        const LINKER      = 1 << 3;
-        /// Sanitizer runtime libraries
-        const SANITIZERS  = 1 << 4;
-        /// Other MinGW libs and Windows import libs
-        const MINGW       = 1 << 5;
-    }
-}
-
-impl FromStr for LinkSelfContainedComponents {
-    type Err = ();
-
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        Ok(match s {
-            "crto" => LinkSelfContainedComponents::CRT_OBJECTS,
-            "libc" => LinkSelfContainedComponents::LIBC,
-            "unwind" => LinkSelfContainedComponents::UNWIND,
-            "linker" => LinkSelfContainedComponents::LINKER,
-            "sanitizers" => LinkSelfContainedComponents::SANITIZERS,
-            "mingw" => LinkSelfContainedComponents::MINGW,
-            _ => return Err(()),
-        })
-    }
+    /// The components that are disabled on the CLI, using the `-component` syntax or one of the
+    /// `false` shortcuts.
+    disabled_components: LinkSelfContainedComponents,
```

Comment:
The setup with `enabled_components` and `disabled_components` is really confusing.
Is it introduced just to continue using bitflags instead of `Option<bool>`s for individual components?

Looks like `-Clink-self-contained=-linker -Clink-self-contained=+linker` is not treated correctly this way.
It will result in both `enabled_components(linker) == true` and `disabled_components(linker) == true`, with `-linker` winning in link.rs.
Which is not correct, because later options override earlier options and `+linker` should win.

I guess the bitflags can be kept and `Option<bool>`s avoided by using 3 sets instead of 2.
```rust
enabled_components: LinkSelfContainedComponents,
disabled_components: LinkSelfContainedComponents,
resulting_components: LinkSelfContainedComponents, // or effective_components, or just components
```
with `(en,dis)abled_components` used only for feature gating (?) and `resulting_components` being used for everything else.

Reply:
> Is it introduced just to continue using bitflags instead of Option<bool>s for individual components?

Yes, somewhat, and just for consistency using `LinkSelfContainedComponents` everywhere; in particular in target specs because it doesn't make much sense for them to *disable* self-contained linking components: components are all disabled by default, target specs only can enable them. Having `Option<bool>`s in `rustc_target` wouldn't be really needed, everything can be modeled with enabled components on the CLI and target specs, and disabled components on the CLI.

The `effective_components` looked less useful to me, as these can only be computed correctly after we've done link-self-contained inference in `link.rs`, by mixing the CLI's components and the target's components, which you can see more fully realized in #116035.

> Looks like `-Clink-self-contained=-linker -Clink-self-contained=+linker` is not treated correctly this way.

I guess that depends on what we want. If it's the last value that wins, then maybe. However, I've continued this in PR https://github.com/rust-lang/rust/pull/116035 where I went with https://github.com/rust-lang/rust/pull/116035/commits/308fa0407849ac7180b803ac88ad6f15a7949667 emitting an error if that happens, and the associated test in https://github.com/rust-lang/rust/pull/116035/commits/b15d81c0723a49d4a2ff145381e97196c897ab26.

I could pull these commits into this PR, or if that's not what we want and would like to allow this CLI combination, implement the later option override.

What do you think?

Reply:
Let's pull that commit to here, and maybe reconsider later when the whole implementation is complete.

Reply:
Agreed.